### ISSUE TITLE
Be able to handle frames of variable size (with known maximum)

### ISSUE THREAD

**Issue by jakirkham** [ilastik/ilastik#1076]
It is often useful to truncate an image due to artifacts on the borders or due to registered frames that stick out. As the shape of the frame cannot be known before computation in these cases, it makes it difficult to work with these in the traditional lazyflow environment.

However, we are fortunate in that the maximum extent is known in advance in all of these cases. As a consequence, we simply need to denote what portion of the frame is useful (or what portion is useless). It turns out that NumPy already supplies this functionality through masked arrays ( http://docs.scipy.org/doc/numpy/reference/maskedarray.generic.html ). These extend ndarrays by supplying a mask to indicate which regions are invalid and a fill value when only an ndarray should be used.

The goal of this issue is to provide this functionality in lazyflow and make sure that this does not affect viewing the ndarray in volumina.

**Comment by jakirkham** [comment]
The data allocation in lazyflow for future results occurs here ( https://github.com/ilastik/lazyflow/blob/master/lazyflow/stype.py#L101 ). If an attribute is added to the metadata about masking, this can be checked at allocation to ensure that a masked array is allocated when needed. Additionally, it is wise to make sure the mask allocated is full size and is not shrunk as it will cause issues when trying to change a view.

**Comment by jakirkham** [comment]
It turns out that volumina generates QImages by using qimage2ndarray ( https://github.com/hmeine/qimage2ndarray ) in the following places for grayscale, alpha modulated, colortable, color, etc ( https://github.com/ilastik/volumina/blob/master/volumina/pixelpipeline/imagesources.py#L201 ) ( https://github.com/ilastik/volumina/blob/master/volumina/pixelpipeline/imagesources.py#L294 ) ( https://github.com/ilastik/volumina/blob/master/volumina/pixelpipeline/imagesources.py#L440 ) ( https://github.com/ilastik/volumina/blob/master/volumina/pixelpipeline/imagesources.py#L529 ) ( https://github.com/ilastik/volumina/blob/master/volumina/pixelpipeline/imagesources.py#L566 ). These all rely on two function from qimage2ndarray. When inspecting the source code of qimage2ndarray, it can be seen that masked arrays are supported all follow the expected behavior of make masked values transparent when using grayscale ( https://github.com/hmeine/qimage2ndarray/blob/master/qimage2ndarray/__init__.py#L243 ) and color ( https://github.com/hmeine/qimage2ndarray/blob/master/qimage2ndarray/__init__.py#L315 ). As a consequence, there is nothing needed for handling viewing the masked arrays. That being said, it might be worth adding a test or two to ensure the support remains.

**Comment by jakirkham** [comment]
Related to ( https://github.com/jakirkham/ilastik/issues/34 ), which has been closed in favor of this.

**Comment by jakirkham** [comment]
Split out PR for layer changes ( https://github.com/ilastik/volumina/pull/153 ).

**Comment by burgerdev** [comment]
As @jakirkham [suggested](https://github.com/ilastik/lazyflow/pull/164#issuecomment-76103999), I will continue posting questions here instead of in the pull request. 

> These extend ndarrays by supplying a mask to indicate which regions are invalid and a fill value when only an ndarray should be used.
> The goal of this issue is to provide this functionality in lazyflow and make sure that this does not affect viewing the ndarray in volumina.

> [...]

>  Having masked arrays as the default sounds dangerous.

I understand your explanation of how a slot gets a masked array and that your pull request solves the original problem of *displaying* masked arrays. What I still don't understand is how a masked array is introduced to the pipeline, ilastik-wise. Is there (or will there be) a cropping operator? Or are there `Applet.topLevelOperator`s that have masked arrays as output slots for the GUI? Is there any possibility that a non-mask-aware slot downstream will get a masked array?

**Comment by jakirkham** [comment]
Thanks. Also, thanks for tying back to my previous comment. @stuarteberg if you would like to follow the discussion, we are over here now.

Before we get started with the questions, I want to clarify a few things.

> I understand your explanation of how a slot gets a masked array and that your pull request solves the original problem of _displaying_ masked arrays.

It is true that we do solve the displaying of masked arrays here ( https://github.com/ilastik/volumina/pull/152 ), but we solved or are working on solutions to a number of other related problems. Namely, allocation ( https://github.com/ilastik/lazyflow/pull/163 ), caching ( https://github.com/ilastik/lazyflow/pull/164 ), and serialization ( https://github.com/ilastik/ilastik/pull/1078 ).

**Comment by jakirkham** [comment]
Excellent questions! Let me see if I am understanding you correctly. There seems to be two main questions, which I have listed below.

1. What sort of operators will be added that always provide a masked array as output?
2. What can a non-mask supporting operator do to ensure it doesn't receive a masked array as input?

&nbsp;
Regarding the first question, there is only __one__ operator that has already been added that would _always_ introduce a masked array. This operator is `OpMaskArray` ( https://github.com/ilastik/lazyflow/commit/06d64827c8a839aed16ccd7c9e3de2b6e57acd06 ). A cropping operator is planned, but has not been included yet. Also, a registration operator is in the works, which will require cropping and thus will fit into this category. There was some discussion about whether some operators that use or apply masks should take advantage of masked arrays ( https://github.com/ilastik/lazyflow/pull/163#issuecomment-73310479 ), but this has not been implemented.

In short, there is only one operator currently in `lazyflow` that you should worry about. A few more will be added in time. As mentioned before ( https://github.com/ilastik/lazyflow/pull/164#issuecomment-76103999 ), a developer could always set up an operator to take a masked array.
&nbsp;
&nbsp;
Regarding the second question, it is pretty simple to guard an operator from receiving a masked array as input. In `setupOutputs`, one can simply assert as shown below. In this example, I have chosen an arbitrary name for input slot, but any relevant input could be used. If there are multiple input that should not have a mask, multiple assertions can be added. Alternatively, a `for` loop over `self.input` is also possible. Of course, the message could be tweaked for the particular case.

    assert not self.SomeInput.meta.has_mask, "The operator, \"%s\", is being setup to recieve a masked array as input to slot, \"%s\". This is currently not supported." % (self.name, self.SomeInput.name)

**Comment by burgerdev** [comment]
Thanks for the nice and well structured explanation. 

>  it is pretty simple to guard an operator from receiving a masked array as input

But there are already loads of operators available, which fall into one of the following categories

  1. I don't care if input is masked or not (`OpArrayPiper` and similar operators)
  2. I am mask-aware (by now the cache operators and some others)
  3. I can make use of the default numpy behaviour (?)
  4. I simply don't know what to do with a masked array, because a mask makes no sense in my context (`OpLabelVolume`, `OpFilterLabels`, vigra filters and prediction operators, ...)

Operators 1-3 cause no problems and can just be used as per usual.
The operators I know of in group 4 would not handle a masked array as you would want them to (i.e. ignore masked region) but would count it as background, large object, strange tumour type, etc. Perhaps it would make sense for operators to explicitly opt-in for masked arrays. Or we would have to check for many existing operators whether they can be used with masked arrays or not. 

Alternatively, we can introduce a new lazyflow user rule:

> Whoever chains two operators is responsible for checking their masked-implementation state.

I think the latter is fine but should be publicly announced at least when the first preprocessing applets become mask-ready.

**Comment by jakirkham** [comment]
> Operators 1-3 cause no problems and can just be used as per usual.

Agreed. In the case of `OpArrayPiper`, I tested it anyway as it is useful to use in other tests and good to verify it works as expected with masked arrays. This was mainly helpful in debugging lazyflow to make sure it supported masked arrays. ( https://github.com/ilastik/lazyflow/commits/08af2d8d56036f58250d2a5087705946ac948e08/tests/testOpArrayPiper.py )

> The operators I know of in group 4 would not handle a masked array as you would want them to...
> Whoever chains two operators is responsible for checking their masked-implementation state.

This rule sounds reasonable and really applies in cases where masks aren't concerned.

However, I would like to go one step further. In the `lazyflow` pull request ( https://github.com/ilastik/lazyflow/pull/164 ), I would like to add the aforementioned assertion to operators in group 4. This way thoughtless chaining of operators doesn't result in mysterious behavior, but an error, which will let the developer know what they have done incorrectly. If it wasn't incorrect and can be made to work, then they could implement the tests necessary to verify that it works and strip the assertion. It may not be a bad idea to add this assertion to operators in group 3, as well. At least until we have verified they do result in the correct behavior via tests.

Though, I can't say I know everywhere that the assertion should be added. So, I'd be inclined to put them in every operator that hasn't been tested. This may not be a bad move as we can always remove the assertions once tests verify that they are working correctly.

**Comment by burgerdev** [comment]
> So, I'd be inclined to put them in every operator that hasn't been tested.

I think this agrees perfectly with what I had in mind for "operators opting in". Something like a magic (class) attribute?

```python
class Operator:
    mask_enabled = False
    # ...

    def _setupOutputs(self):
        if not self.mask_enabled:
            for s in self.inputs:
                assert not s.meta.has_mask
        # ...


class OpArrayCache(...):
    mask_enabled = True
    # ...

class OpCompressedCache(...):
    mask_enabled = True
    # ...

# ...
```

**Comment by jakirkham** [comment]
It does indeed sound like we agree about protecting existing operators. Also, I do appreciate your enthusiasm. I'd just like to point out that the choice to support masked arrays is a per slot not per operator decision. So, it should be some sort of attribute that determines whether the slot will support a mask.

For instance, take `OpArrayCache`. We have supported masked arrays provided to `Input`. However, it is unclear to me what providing a masked array to `blockShape` should mean or whether it will ever have any meaning.

    class OpArrayCache(OpCache):
        ...

        #Input
        Input = InputSlot()
        blockShape = InputSlot(value = DefaultBlockSize)
        fixAtCurrent = InputSlot(value = False)
       
        #Output
        CleanBlocks = OutputSlot()
        Output = OutputSlot()

As to error versus assertion, I'm not too picky here. I only used an assertion as this is a programmer mistake not some unexpected surprise that should be handled.

I'll write up something to add to lazyflow ( https://github.com/ilastik/lazyflow/pull/164 ) and we can debate the details over there. Also, I expect @stuarteberg will have something to say about this, as well.

**Comment by jakirkham** [comment]
@burgerdev feel free to take a look at the pull request again ( https://github.com/ilastik/lazyflow/pull/164 ). I have provided an `allow_mask` keyword argument to `Slot`, which defaults to `False`. When `False` this results in `AssertionError`s if the value of the slot is set to a masked array, if a masked array is provided via `setInSlot`, or if another slot is connected, whose `has_mask` value is `True`. In addition, I have provided tests to verify that this error is raised when needed and not when it should not occur.
